pkg/webhook/registry/validating.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2020 Clastix Labs.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registry
+
+import (
+	"context"
+	"net/http"
+	"regexp"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	capsulev1alpha1 "github.com/clastix/capsule/api/v1alpha1"
+	capsulewebhook "github.com/clastix/capsule/pkg/webhook"
+)
+
+// +kubebuilder:webhook:path=/validating-v1-registry,mutating=false,failurePolicy=fail,groups="",resources=pods,verbs=create,versions=v1,name=pod.capsule.clastix.io
+
+type webhook struct {
+	handler capsulewebhook.Handler
+}
+
+func Webhook(handler capsulewebhook.Handler) capsulewebhook.Webhook {
+	return &webhook{handler: handler}
+}
+
+func (w *webhook) GetName() string {
+	return "registry"
+}
+
+func (w *webhook) GetPath() string {
+	return "/validating-v1-registry"
+}
+
+func (w *webhook) GetHandler() capsulewebhook.Handler {
+	return w.handler
+}
+
+type handler struct {
+}
+
+func Handler() capsulewebhook.Handler {
+	return &handler{}
+}
+
+func (h *handler) OnCreate(c client.Client, decoder *admission.Decoder) capsulewebhook.Func {
+	return func(ctx context.Context, req admission.Request) admission.Response {
+		var valid, matched bool
+		pod := &v1.Pod{}
+
+		if err := decoder.Decode(req, pod); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		containers := pod.Spec.Containers
+
+		for _, container := range containers {
+			if container.Image == "" {
+				return admission.Errored(http.StatusBadRequest, NewRegistryClassNotValid())
+			}
+		}
+
+		tl := &capsulev1alpha1.TenantList{}
+		if err := c.List(ctx, tl, client.MatchingFieldsSelector{
+			Selector: fields.OneTermEqualSelector(".status.namespaces", pod.Namespace),
+		}); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		if len(tl.Items[0].Spec.RegistryClasses.Allowed) > 0 {
+			for _, container := range containers {
+				valid = tl.Items[0].Spec.RegistryClasses.Allowed.IsStringInList(container.Image)
+			}
+		}
+
+		if len(tl.Items[0].Spec.RegistryClasses.Allowed) > 0 {
+			for _, container := range containers {
+				matched, _ = regexp.MatchString(tl.Items[0].Spec.RegistryClasses.AllowedRegex, container.Image)
+			}
+		}
+
+		if !valid && !matched {
+			for _, container := range containers {
```

Comment:
I guess we can easily avoid to iterate over all the containers, here.

If not valid and not matched, let's disallow it.

Reply:
same reason 
> A pod can have multiple containers each having multiple images, so the webhook must check each image in case they are from a trusted registery. Thats why we need to range over all containers

Reply:
There's no need to iterate here: we can safely return the admission error without iteration, here.

Reply:
In case i remove range, is this the error msg you want to go with ? ```NewregistryClassForbidden(pod.Spec.Containers[0].Image)```

Reply:
BTW can we use admissioned.denied here ?

Reply:
Would be nice adding the list of container images that are not allowed to run using the `admission.Denied`, something like:

```
The Pod cannot be created due to following image(s) not allowed: ${images...}
```

If it's too difficult to collect a slice of unmatched and non valid images, let's ignore this: let's focus on addressing the feature.